File: server/fakes/state_inspector.go
// screenreader-mcp fakes -- FakeStateInspector: the StateInspector port double.
// Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
//
// ROLE: test double for domain/ports/state_inspector.go.
// USED BY: the get_state tool controller tests.
package fakes

import (
	"sync"

	"github.com/marlon-sousa/screen-readers-mcp/server/domain/ports"
)

type FakeStateInspector struct {
	mu    sync.Mutex
	state ports.ReaderState
	err   error
}

var _ ports.StateInspector = (*FakeStateInspector)(nil)

func NewFakeStateInspector() *FakeStateInspector { return &FakeStateInspector{} }

func (f *FakeStateInspector) SetState(state ports.ReaderState) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.state = state
}

func (f *FakeStateInspector) FailWith(err error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.err = err
}

func (f *FakeStateInspector) State() (ports.ReaderState, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	if f.err != nil {
		return ports.ReaderState{}, f.err
	}
	return f.state, nil
}

File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/ports/state_inspector.py
# nvdaMcpBridge domain -- the StateInspector port.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: domain port; reads the reader's mode-state: browse mode, speech mode, sleep mode and input help.
# USED BY: the GetStateHandler.
# IMPLEMENTED BY: adapters/nvda_state_inspector.py and tests/fakes/state_inspector.py.
# browse_mode "none" means the focus has no TreeInterceptor, which is not the same as "focus".

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass


@dataclass(frozen=True)
class ReaderState:
	browse_mode: str  # "browse" | "focus" | "none"
	speech_mode: str  # "talk" | "off" | "beeps" | "onDemand"
	sleep_mode: bool
	input_help: bool


class StateInspector(ABC):
	@abstractmethod
	def state(self) -> ReaderState:
		pass
